engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java

```diff
@@ -3208,10 +3206,50 @@ public Object call() throws Exception
     return null;
   }
 
+  private class ChangeDAGRunnable implements Callable<Object>
+  {
+    private final transient DAGChangeSetImpl request;
+
+    public ChangeDAGRunnable(DAG.DAGChangeSet changes)
+    {
+      request = (DAGChangeSetImpl)changes;
+    }
+
+    @Override
+    public Object call() throws Exception
+    {
+      if (request == null) {
+        return null;
+      }
+
+      try {
+        LOG.info("starting plan change ");
+        LogicalPlan lp = plan.cloneLogicalPlan();
+        PlanModifier pm = new PlanModifier(lp);
+        pm.applyDagChangeSet(request);
+        lp.validate();
+
+        LOG.info("validated logical plan, now making actual change");
+        pm = new PlanModifier(plan);
+        pm.applyDagChangeSet(request);
+        LOG.info("plan change done");
+        return null;
+      } finally {
+        plan.clearPendingChangeRequest();
```

Comment:
At a time only one pending dag change request will be active. once the dag change request is processed successfully or unsuccessfully, it will be cleared so that new request can be submitted from statsListener. Stats listener has access to the feature through with it can check the status of last submitted dag change request.